solver/src/settlement_submission/archer_settlement.rs

```diff
@@ -0,0 +1,377 @@
+// so that the solver can run again.
+// In addition to simulation failure we make use of a deadline after which submission attempts also
+// stop. This allows the solver to update and improve a solution even if it hasn't yet become
+// invalid.
+// We do not know in advance which of our submitted transactions will get mined. Instead of polling
+// all of them we only check the account's nonce as an optimization. When this happens all our
+// transactions definitely become invalid (even if the transaction came for whatever reason
+// from outside) so it is only at that point that we need to check the hashes individually to the
+// find the one that got mined (if any).
+
+// Idea:
+// The current driver code is either solving or waiting for a solution to be mined.
+// When using archer we can improve on this by always solving the current order book as if there was
+// no pending transaction and continually updating the transaction we are sending to archer.
+// This is a bigger change so the code here still adheres to the traditional tx submission
+// workflow.
+
+// TODO: Make sure settlement contract always has eth to pay the miner. Could reuse buffers or just
+// manually fund it like we do with other solver accounts and make part of our monitor scripts.
+
+// TODO: Node gas estimates seem a little higher than actual. This can be seen when comparing
+// to tenderly simulations and mined transactions. This causes us to overpay somewhat.
+
+use super::{archer_api::ArcherApi, ESTIMATE_GAS_LIMIT_FACTOR};
+use crate::{interactions::block_coinbase, settlement::Settlement};
+use anyhow::{anyhow, Context, Result};
+use chrono::{DateTime, Utc};
+use contracts::GPv2Settlement;
+use ethcontract::{errors::MethodError, transaction::Transaction, Account, GasPrice};
+use futures::FutureExt;
+use gas_estimation::GasPriceEstimating;
+use primitive_types::{H256, U256};
+use shared::Web3;
+use std::{
+    sync::Arc,
+    time::{Duration, Instant, SystemTime},
+};
+use web3::types::TransactionId;
+
+pub struct ArcherSolutionSubmitter {
+    pub web3: Web3,
+    pub contract: GPv2Settlement,
+    pub account: Account,
+    pub archer_api: ArcherApi,
+    pub gas_price_estimator: Arc<dyn GasPriceEstimating>,
+    pub gas_price_cap: f64,
+}
+
+impl ArcherSolutionSubmitter {
+    /// Submit a settlement to the contract, updating the transaction with gas prices if they increase.
+    ///
+    /// Goes through the archerdao network so that failing transactions do not get mined and thus do
+    /// not cost gas.
+    ///
+    /// Returns None if the deadline is reached without a mined transaction.
+    ///
+    /// Only works on mainnet.
+    pub async fn submit(
+        &self,
+        target_confirm_time: Duration,
+        deadline: SystemTime,
+        settlement: Settlement,
+        gas_estimate: U256,
+    ) -> Result<Option<H256>> {
+        let nonce = self.nonce().await?;
```

Comment:
I don't plan on mixing them for the same settlement (but it could be interesting) so this should be fine. If we did do that I guess it would still be fine because the settling the new transaction should fail simulation and as you said not be taken by miners.